src/Microsoft.SqlTools.ServiceLayer/Utility/SqlScriptFormatters/FromSqlScript.cs

```diff
@@ -0,0 +1,148 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.SqlTools.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer.Utility.SqlScriptFormatters
+{
+    /// <summary>
+    /// Provides utilities for converting from SQL script syntax into POCOs.
+    /// </summary>
+    public static class FromSqlScript
+    {
+        private static readonly Regex StringRegex = new Regex("^N?'(.*)'$", RegexOptions.Compiled);
+        
+        /// <summary>
+        /// Decodes a multipart identifier as used in a SQL script into an array of the multiple
+        /// parts of the identifier. Implemented as a state machine that iterates over the
+        /// characters of the multipart identifier.
+        /// </summary>
+        /// <param name="multipartIdentifier">Multipart identifier to decode (eg, "[dbo].[test]")</param>
+        /// <returns>The parts of the multipart identifier in an array (eg, "dbo", "test")</returns>
+        /// <exception cref="FormatException">
+        /// Thrown if an invalid state transition is made, indicating that the multipart identifer
+        /// is not valid. 
+        /// </exception>
+        public static string[] DecodeMultipartIdenfitier(string multipartIdentifier)
```

Comment:
Don't we have utility methods that already do this? We certainly have ones in the DacFx codebase. There are a number of issues here:
- Whether to handle quoted identifiers or not
- Detection of escaped `[` and `]` characters, which isn't handled
- Potentially others. Quoted identifier handling is an area where I've seen so many bugs raised in the past, that the only "safe" way to handle is to reuse existing code or waste lots of time.

I'm going to wait on the PR and sync with you on where to get resusable code from. There's almost certainly a util method in our existing code that we reference / have ported over.

Reply:
OK looking further... I realize you already wrote this (and it's already shipped this way). Checking with Chris, we could use `Microsoft.SqlServer.SqlParser/Parser/EscapeSequence.cs#833153` for these. It covers unescaping both escaped characters and escaped literals. If possible, could we look at using this since it's battle hardened?

Reply:
I suppose that's fine, but didn't we already try using SqlParser and then got rid of it because it was buggy?